### ISSUE TITLE
Retroachivements menu

### ISSUE THREAD

**Issue by ethanaobrien** [emulatorjs/emulatorjs#589]
If the game supports it, add a retroarchivements button to the menu bar at the bottom

**Comment by michael-j-green** [comment]
Probably on your radar for this already, it would be awesome if we could pass the retroachievements API key via an EJS_ var.

**Comment by ethanaobrien** [comment]
Enabling RetroAchivements includes setting `HAVE_CHEEVOS` to `0` in the emscripten makefile. Dont know how easy this will be

**Comment by GavriloviciEduard** [comment]
Any updates on this?

**Comment by ethanaobrien** [comment]
No updates. This is a difficult task and is not a priority

**Comment by michael-j-green** [comment]
This one might be tricky, if not impossible.

The Retro Achievements site has strict requirements for connectivity to prevent cheating, and EJS runs in the browser which imposes some pretty strict security especially around network connectivity - for obvious reasons.

**Comment by ethanaobrien** [comment]
We would likely need a proxy server dedicated for this

**Comment by michael-j-green** [comment]
This was my thought too... since WebAssembly can only handle WebSockets, you'd need something to translate that to a real TCP call.

This might be a feature that projects that enclose EJS would have to implement, and not EJS (other than providing a way to set the proxy address).

But the problem becomes: RA doesn't publish their API for submitting achievements (as far as I can find), and while we could ascertain their endpoints with WireShark or similar, RA would still frown upon it and potentially ban you for attempting to use a "hacked" client.

While I LOVE the idea, I'm thinking this is going to be a massive amount of work with little gain, with the potential for it all to be thrown out because of RA (which from their point of view would be understandable). I do wish RA could publish their API so that RetroArch could be pointed at an alternative site (like a project that includes EJS).

**Comment by michael-j-green** [comment]
Of course right after I comment about not being able to find something, I find it.

It's all in the retro achievements repo.

Fact remains though that it's a large engineering effort.

**Comment by michael-j-green** [comment]
@ethanaobrien; looks like there is already a proxy out there https://github.com/novnc/websockify

But it does bring to light issues with SSL encryption which doesn't sound exactly end user friendly: https://github.com/novnc/websockify#encrypted-websocket-connections-wss (assuming that this part is relevant which I can only assume that it would be).

**Comment by ethanaobrien** [comment]
> since WebAssembly can only handle WebSockets, you'd need something to translate that to a real TCP call.

I know ways around this actually, pretty simple

> This might be a feature that projects that enclose EJS would have to implement, and not EJS (other than providing a way to set the proxy address).

Correct 

> RA doesn't publish their API for submitting achievements (as far as I can find), and while we could ascertain their endpoints with WireShark or similar, RA would still frown upon it and potentially ban you for attempting to use a "hacked" client.

Maybe we should check the terms and conditions to see if this is something we can even do without violating those?

> While I LOVE the idea, I'm thinking this is going to be a massive amount of work with little gain, with the potential for it all to be thrown out because of RA (which from their point of view would be understandable).

Which is why I haven't looked into this at all really

> Of course right after I comment about not being able to find something, I find it.
It's all in the retro achievements repo.

Oh so you did find it, nice. Though that's still a lot of effort for something so little

> But it does bring to light issues with SSL encryption which doesn't sound exactly end user friendly

Looks like just a simple cert and key is needed, standard https encryption. With the way they worded it, it doesn't seem mandatory.

**Comment by michael-j-green** [comment]
I think we come back to classifying this as low priority, with a significant amount of work to implement.

We’ll leave the issue here to track it because it’s a cool feature, but no one has the time to implement it or work with the RA team to ensure it won’t get booted.

**Comment by mikeyjoel** [comment]
Would love to see this implemented. I'll take a look into it and see what I can do. Got my romm manifest on kubeadm + argocd  running and this is the last piece missing...

**Comment by virtualducttape** [comment]
Agreed, would love to see this implemented. Given that https://github.com/misantronic/RAOfflineProxy is officially approved for use by RetroAchievements and it has more (and arguably way "shadier") hoops it has to jump through to work offline, it seems likely that EmulatorJS could build a version that RA was fine with. 

FWIW, my use case is using ROMM, I'm able to play the same save file on portable devices or through the browser with EmulatorJS, I'd love to have my achievements still count no matter which emulator I'm playing from.

Thanks for all your work, it's wonderful to have a browser-based emulator at all, nevermind one that's as terrific as this one!
